src/scheduler.rs

```diff
@@ -0,0 +1,1013 @@
+        preferred_nodes: &[Option<&NodeIdentity>],
+    ) -> Vec<Option<NodeIdentity>>;
+}
+
+/// Implements a pod placement strategy where no two pods from the same role+group
+/// are scheduled on the same node at the same time. It fails if there are not enough nodes to place pods on.
+/// This useful for when pods are deployed on a bare metal K8S environment with Stackable agents as nodes.
+struct GroupAntiAffinityStrategy<'a> {
+    eligible_nodes: RefCell<RoleGroupEligibleNodes>,
+    existing_mapping: &'a PodToNodeMapping,
+}
+
+/// Implements a pod placement strategy where pods are hashed to eligible nodes without regards to
+/// the existing mapping. This useful for when pods are deployed as containers on a standard K8S
+/// environment.
+struct HashingStrategy<'a> {
+    eligible_nodes: &'a RoleGroupEligibleNodes,
+    hasher: RefCell<DefaultHasher>,
+}
+
+pub enum ScheduleStrategy {
+    /// A scheduling strategy that will refuse to schedule two pods within one role+group on the same
+    /// node. If no enough pods are available, the pod will not be scheduled on any node.
+    /// This useful for when pods are deployed on a bare metal K8S environment with Stackable agents as nodes.
+    GroupAntiAffinity,
+    /// A scheduling strategy that will simply hash the pod onto one of the existing pods without
+    /// any consideration for the distribution of all other existing pods.This useful for when pods
+    /// are deployed as containers on a standard K8S environment.
+    Hashing,
+}
+
+/// A scheduler implementation that remembers where pods were once scheduled (based on
+/// their ids) and maps them to the same nodes in the future. The `history` provides preferred
+/// nodes to map onto based past mappings.
+/// The `strategy` might choose a different node if the history node cannot be used.
+pub struct StickyScheduler<'a, H: PodPlacementHistory> {
+    pub history: &'a mut H,
+    pub strategy: ScheduleStrategy,
+}
+
+//--------------------------------------------------------------------------------
+// Implementation
+//--------------------------------------------------------------------------------
+
+impl<'a> K8SUnboundedHistory<'a> {
+    pub fn new(client: &'a Client, history: PodToNodeMapping) -> Self {
+        K8SUnboundedHistory {
+            client,
+            history,
+            modified: false,
+        }
+    }
+
+    pub async fn save<T>(&mut self, resource: &T) -> OperatorResult<T>
+    where
+        T: Clone + Debug + DeserializeOwned + Resource,
+        <T as Resource>::DynamicType: Default,
+    {
+        if self.modified {
+            return match self
+                .client
+                .merge_patch_status(resource, &json!({ "history": self.history }))
+                .await
+            {
+                Ok(res) => {
+                    self.modified = false;
+                    Ok(res)
+                }
+                err => err,
+            };
+        }
+
+        Ok(resource.clone())
```

Comment:
I have a similar problem with the versioning. I dont like the clone everytime when basically nothing happens. 

Ill update here once i figured it out, or maybe @siegfriedweber has an idea.

We cannot return a reference here, i think we should just mut& T the resource?

Reply:
You could return a `std::borrow::Cow`. But I am not sure if the extra complexity is worth it.

Reply:
We decided to leave it as is for now.